builder/bundles.go

```
@@ -480,30 +515,149 @@ func downloadRpms(packagerCmd, rpmList []string, baseDir string, maxRetries int)
-		for p := range bundle.AllPackages {
-			rpmList = append(rpmList, p)
+	err = rpm2Cmd.Start()
+	if err != nil {
+		return fmt.Errorf("rpm2cpio failed with %v", err)
+	}
+	err = cpioCmd.Start()
+	if err != nil {
+		return fmt.Errorf("cpio failed with %v", err)
+	}
+
+	err = rpm2Cmd.Wait()
+	if err != nil {
+		return fmt.Errorf("rpm2cpio failed with %v", err)
+	}
+
+	err = pw.Close()
+	if err != nil {
+		return fmt.Errorf("  writer pipe failed to close:  %v ", err)
+	}
+
+	err = cpioCmd.Wait()
+	if err != nil {
+		return fmt.Errorf("cpioCmd failed with %v", err)
+	}
+
+	err = pr.Close()
+	if err != nil {
+		return fmt.Errorf(" reader pipe failed to close: %v ", err)
+	}
+
+	return err
+}
+
+func installBundleToFull(packagerCmd []string, baseDir string, bundle *bundle, downloadRetries int, numWorkers int, localPath string) error {
+	var err error
+	var wg sync.WaitGroup
+	rpmCh := make(chan string)
+
+	var errCh error
+	var rpmList []string
+	var rpmPath string
+
+	if len(bundle.AllRpmPackages) < numWorkers {
+		numWorkers = len(bundle.AllRpmPackages)
+	}
+	errorCh := make(chan error, numWorkers)
+	wg.Add(numWorkers)
+
+	rpmWorker := func() {
+		var e error
+		for rpm := range rpmCh {
+			for i := 0; i < extractRetries; i++ {
+				e = extractRpm(baseDir, rpm)
+				if e == nil {
+					break
+				}
+			}
+			if e != nil {
+				errorCh <- e
+				break
+			}
 		}
+		wg.Done()
+	}
 
-		// Retry RPM downloads to avoid timeout failures due to slow network
-		_, err = downloadRpms(packagerCmd, rpmList, baseDir, downloadRetries)
-		if err != nil {
-			return err
+	for i := 0; i < numWorkers; i++ {
+		go rpmWorker()
+	}
+
+	if !Offline {
+		for rpm := range bundle.AllRpmPackages {
+			rpmFull := filepath.Join(localPath, rpm)
+			if _, err = os.Stat(rpmFull); os.IsNotExist(err) {
+				rpm1 := strings.TrimSuffix(rpm, ".rpm")
+				rpmList = append(rpmList, rpm1)
+			}
+		}
+		if len(rpmList) > 0 {
+			rpmPath, err = downloadRpm(packagerCmd, rpmList, baseDir, downloadRetries)
```

this should be inside an `if !Offline` block and have an else that returns error. This way, if there is any missing rpm, we fail early.